File: jimsfridge/mqtttoazure/StopWatch.py
#----------------------------------------------------------------------------
#
#  $Workfile: StopWatch.py$
#
#  $Revision: 126$
#
#  Project:    JimsFridge
#
#                            Copyright (c) 2017
#                               Jim Wright
#                            All Rights Reserved
#
#  Modification History:
#  $Log:
#  $
#
#----------------------------------------------------------------------------

#----------------------------------------------------------------------------
#  Imports
#----------------------------------------------------------------------------
import time

#----------------------------------------------------------------------------
#  Class Declarations
#----------------------------------------------------------------------------
#
# Class Name: IconBuilder
# 
# Purpose:
#      Handles the main form
#
#----------------------------------------------------------------------------
class StopWatch:

  #--------------------------------------------------------------------
  # Purpose:
  #     Constructor
  #
  # Notes:
  #     None.
  #--------------------------------------------------------------------
  def __init__(self, waitTime=1000, isSeconds=False):
    self.mLastTime = 0
    self.mWaitTime = waitTime
    self.mIsSeconds = isSeconds

  #--------------------------------------------------------------------
  # Purpose:
  #     Change the wait time
  #
  # Notes:
  #     None.
  #--------------------------------------------------------------------
  def setTime(self, waitTime):
    self.mWaitTime = waitTime

  #--------------------------------------------------------------------
  # Purpose:
  #     Return time in ms
  #
  # Notes:
  #     None.
  #--------------------------------------------------------------------
  @staticmethod
  def now():
    return int(round(time.time() * 1000))

  #--------------------------------------------------------------------
  # Purpose:
  #     Return time in seconds
  #
  # Notes:
  #     None.
  #--------------------------------------------------------------------
  @staticmethod
  def nowInSeconds():
    return int(round(time.time()))

  #--------------------------------------------------------------------
  # Purpose:
  #     Has the stopwatch expired
  #
  # Notes:
  #     None.
  #--------------------------------------------------------------------
  def isExpired(self):
    if (False == self.mIsSeconds):
      if ((StopWatch.now() - self.mLastTime) > self.mWaitTime):
        return True
    else:
      if ((StopWatch.nowInSeconds() - self.mLastTime) > self.mWaitTime):
        return True
    return False

  #--------------------------------------------------------------------
  # Purpose:
  #     Reset stopwatch
  #
  # Notes:
  #     None.
  #--------------------------------------------------------------------
  def reset(self):
    if (False == self.mIsSeconds):
      self.mLastTime = StopWatch.now()
    else:
      self.mLastTime = StopWatch.nowInSeconds()

  #--------------------------------------------------------------------
  # Purpose:
  #     Return time left
  #
  # Notes:
  #     None.
  #--------------------------------------------------------------------
  def getTimeLeft(self):
    if (False == self.mIsSeconds):
      return self.mWaitTime - (StopWatch.now() - self.mLastTime)
    else:
      return self.mWaitTime - (StopWatch.nowInSeconds() - self.mLastTime)

File: jimsfridge/mqtttoazureCPP/StopWatch.cpp
//----------------------------------------------------------------------------
//
//  $Workfile: StopWatch.cpp$
//
//  $Revision: X$
//
//  Project:    JimsFridge
//
//                            Copyright (c) 2017
//                                Jim Wright
//                            All Rights Reserved
//
//  Modification History:
//  $Log:
//  $
//
//----------------------------------------------------------------------------
#include <stdio.h>
#include <sys/time.h>
#include "StopWatch.h"
 
StopWatch::StopWatch() :
  mLastTime(0),
  mWaitTime(1000),
  mIsSeconds(false)
{
}

StopWatch::StopWatch(uint32_t waitTime) :
  mLastTime(0),
  mWaitTime(waitTime),
  mIsSeconds(false)
{
}

StopWatch::StopWatch(bool isSeconds) :
  mLastTime(0),
  mWaitTime(1000),
  mIsSeconds(isSeconds)
{
}

StopWatch::StopWatch(uint32_t waitTime, bool isSeconds) :
  mLastTime(0),
  mWaitTime(waitTime),
  mIsSeconds(isSeconds)
{
}

uint32_t StopWatch::Now(void)
{
  struct timeval theTime;
  gettimeofday(&theTime, NULL);
  return ((theTime.tv_sec * 1000) + (theTime.tv_usec / 1000));
}

uint32_t StopWatch::NowInSeconds(void)
{
  struct timeval theTime;
  gettimeofday(&theTime, NULL);
  return theTime.tv_sec;
}

void StopWatch::SetTime(uint32_t waitTime)
{
  mWaitTime = waitTime;
}

bool StopWatch::IsExpired(void)
{
  if (false == mIsSeconds)
  {
    if ((Now() - mLastTime) > mWaitTime)
    {
      return true;
    }
  }
  else
  {
    if ((NowInSeconds() - mLastTime) > mWaitTime)
    {
      return true;
    }
  }
  return false;
}

void StopWatch::Reset(void)
{
  if (false == mIsSeconds)
  {
    mLastTime = Now();
  }
  else
  {
    mLastTime = NowInSeconds();
  }
}

int32_t StopWatch::GetTimeLeft()
{
  if (false == mIsSeconds)
  {
    return mWaitTime - (Now() - mLastTime);
  }
  else
  {
    return mWaitTime - (NowInSeconds() - mLastTime);
  }
}

